python/restate/harness.py

```diff
@@ -0,0 +1,223 @@
+                raise e
+            finally:
+                self.exit_event.set()
+
+        self.thread = run_in_background(run_asgi())
+        return self
+
+class HostProxyContainer(DockerContainer):
+    """create a proxy container that proxies a unix domain socket from the host"""
+
+    def __init__(self, name, port, host_uds,  network):
+        super().__init__("alpine/socat")
+        self.with_name(name)
+        self.with_exposed_ports(port)
+        self.with_network(network)
+        self.with_volume_mapping(host_uds, "/tmp/unix_socket", mode="rw")
+        self.with_command(["TCP-LISTEN:9081,reuseaddr,fork", "UNIX-CLIENT:/tmp/unix_socket"])
+        self.name = name
+        self.port = port
+
+    def connection_string(self):
+        """return the connection string for the proxy"""
+        return f"http://{self.name}:{self.port}"
+
+class RestateContainer(DockerContainer):
+    """Create a Restate container"""
+
+    def __init__(self, image, network):
+        super().__init__(image)
+        self.with_exposed_ports(8080, 9070)
+        self.with_network(network)
+
+    def ingress_url(self):
+        """return the URL to access the Restate ingress"""
+        return f"http://{self.get_container_host_ip()}:{self.get_exposed_port(8080)}"
+
+    def admin_url(self):
+        """return the URL to access the Restate admin"""
+        return f"http://{self.get_container_host_ip()}:{self.get_exposed_port(9070)}"
+
+    def get_admin_client(self):
+        """return an httpx client to access the admin interface"""
+        return httpx.Client(base_url=self.admin_url())
+
+    def get_ingress_client(self):
+        """return an httpx client to access the ingress interface"""
+        return httpx.Client(base_url=self.ingress_url())
+
+    @wait_container_is_ready(httpx.HTTPError, URLError)
+    def _wait_healthy(self):
+        """wait for restate's health checks to pass"""
+        self.get_ingress_client().get("/restate/health").raise_for_status()
+        self.get_admin_client().get("/health").raise_for_status()
+
+
+    def start(self):
+        """start the container and wait for health checks to pass"""
+        super().start()
+        self._wait_healthy()
+        return self
+
+
+
+class RestateTestHarness:
+    """A test harness for running Restate SDKs"""
+    uds_path: typing.Optional[str] = None
+    network: typing.Optional[Network] = None
+    server: typing.Optional[UdsAsgiServer] = None
+    restate: typing.Optional[RestateContainer] = None
+    proxy: typing.Optional[HostProxyContainer] = None
+
+
+    def __init__(self, asgi_app):
+        self.asgi_app = asgi_app
+
+    def start(self):
+        """start the restate server and the sdk"""
+        self.uds_path = f"/tmp/restate-test-{random.randint(1, 1 << 63)}.sock"
+        self.server = UdsAsgiServer(self.asgi_app, self.uds_path).start()
+
+        self.network = Network().create()
+
+        self.restate = RestateContainer(image="restatedev/restate:1.1",
```

Comment:
I think this should be latest, otherwise we need to bump the python sdk every time we update the runtime. Plus i find counterintuitive that we decide here to pin to some version X.

Let the user override the restate image though, so everyone's happy and can pin to a specific restate version if they need to.